src/c++/vdo/base/message-stats.c

```diff
@@ -430,3 +432,59 @@ int vdo_write_stats(struct vdo *vdo, char *buf, unsigned int maxlen)
 	vdo_free(stats);
 	return VDO_SUCCESS;
 }
+
+static void write_index_memory(char *prefix, u32 mem, char *suffix,
+			       char **buf, unsigned int *maxlen)
+{
+	/* Determine if the index memory value is fractional */
+	if (mem > 1024) {
+		/* Convert index memory to fractional value */
+		if (mem == (u32)UDS_MEMORY_CONFIG_256MB)
+			write_string(prefix, "0.25", suffix, buf, maxlen);
+		else if (mem == (u32)UDS_MEMORY_CONFIG_512MB)
+			write_string(prefix, "0.50", suffix, buf, maxlen);
+		else
+			write_string(prefix, "0.75", suffix, buf, maxlen);
+	}
+	write_u32(prefix, mem, suffix, buf, maxlen);
+}
+
+static void write_index_config(char *prefix, struct vdo *vdo, char *suffix,
+			       char **buf, unsigned int *maxlen)
+{
+	struct index_config *config = &vdo->geometry.index_config;
+
+	write_string(prefix, "{ ", NULL, buf, maxlen);
+	/* index mem size */
+	write_index_memory("memorySize : ", config->mem, ", ", buf, maxlen);
+	/* whether the index is sparse or not */
+	write_bool("isSparse : ", config->sparse, ", ", buf, maxlen);
+	write_string(NULL, "}", suffix, buf, maxlen);
+}
+
+
+static void write_vdo_config(struct vdo *vdo, char **buf, unsigned int *maxlen)
+{
+	struct vdo_config *config = &vdo->states.vdo.config;
+
+	write_string(NULL, "{ ", NULL, buf, maxlen);
+	/* version */
+	write_u32("version : ", 1, ", ", buf, maxlen);
+	/* physical size */
+	write_block_count_t("physicalSize : ", config->physical_blocks * VDO_BLOCK_SIZE, ", ",
+			    buf, maxlen);
+	/* logical size */
+	write_block_count_t("logicalSize : ", config->logical_blocks * VDO_BLOCK_SIZE, ", ",
+			    buf, maxlen);
+	/* slab size */
+	write_block_count_t("slabSize : ", config->slab_size, ", ", buf, maxlen);
+	/* index config */
+	write_index_config("index : ", vdo, ", ", buf, maxlen);
+	write_string(NULL, "}", NULL, buf, maxlen);
```

Comment:
Is there a good reason to make this a nested structure? It might be simple just to put the two index values at the same level as everything else. Also, I'm not sure this needs three functions to write out the config. The existing code is written this way in order to make the code generation simpler but we don't need to cater to that any more.

Reply:
Its useful to separate it like this, since index: is a sub item in the yaml. This allows it to be index: {  foo }. We could maybe merge everything together, but I much prefer it to be separate.